File: spec/java/src/io/kaitai/struct/spec/TestSwitchManualStr.java
package io.kaitai.struct.spec;

import io.kaitai.struct.testformats.SwitchManualStr;
import io.kaitai.struct.testformats.SwitchManualStr.Opcode.Intval;
import io.kaitai.struct.testformats.SwitchManualStr.Opcode.Strval;
import org.testng.annotations.Test;

import static org.testng.Assert.assertEquals;

public class TestSwitchManualStr extends CommonSpec {
    @Test
    public void testSwitchManualStr() throws Exception {
        SwitchManualStr r = SwitchManualStr.fromFile(SRC_DIR + "switch_opcodes.bin");

        assertEquals(r.opcodes().size(), 4);

        assertEquals(r.opcodes().get(0).code(), "S");
        assertEquals(((Strval) r.opcodes().get(0).body()).value(), "foobar");

        assertEquals(r.opcodes().get(1).code(), "I");
        assertEquals(((Intval) r.opcodes().get(1).body()).value(), 0x42);

        assertEquals(r.opcodes().get(2).code(), "I");
        assertEquals(((Intval) r.opcodes().get(2).body()).value(), 0x37);

        assertEquals(r.opcodes().get(3).code(), "S");
        assertEquals(((Strval) r.opcodes().get(3).body()).value(), "");
    }
}

File: spec/rust/tests/test_switch_manual_str.rs
#![allow(dead_code)]
use std::{fs, rc::Rc};

extern crate kaitai;
use self::kaitai::*;
mod formats;
use formats::switch_manual_str::*;

#[test]
fn test_switch_manual_str() {
    let bytes = fs::read("../../src/switch_opcodes.bin").unwrap();
    let _io = BytesReader::new(&bytes);
    let res = SwitchManualStr::read_into(&_io, None, None);
    let r : Rc<SwitchManualStr>;

    if let Err(err) = res {
        panic!("{:?}", err);
    } else {
        r = res.unwrap();
    }
    assert_eq!(4, r.opcodes().len());

    assert_eq!("S", *r.opcodes()[0].code());
    if let Some(SwitchManualStr_Opcode_Body::SwitchManualStr_Opcode_Strval(s)) = r.opcodes()[0].body().as_ref() {
        assert_eq!("foobar", *s.value());
    } else {
        panic!("expected enum SwitchManualStr_Opcode_Strval");
    }

    assert_eq!("I", *r.opcodes()[1].code());
    if let Some(SwitchManualStr_Opcode_Body::SwitchManualStr_Opcode_Intval(s)) = r.opcodes()[1].body().as_ref() {
        assert_eq!(66, *s.value());
    } else {
        panic!("expected enum SwitchManualStr_Opcode_Intval");
    }

    assert_eq!("I", *r.opcodes()[2].code());
    if let Some(SwitchManualStr_Opcode_Body::SwitchManualStr_Opcode_Intval(s)) = r.opcodes()[2].body().as_ref() {
    assert_eq!(55, *s.value());
    } else {
        panic!("expected enum SwitchManualStr_Opcode_Intval");
    }

    assert_eq!("S", *r.opcodes()[3].code());
    if let Some(SwitchManualStr_Opcode_Body::SwitchManualStr_Opcode_Strval(s)) = r.opcodes()[3].body().as_ref() {
        assert_eq!("", *s.value());
    } else {
        panic!("expected enum SwitchManualStr_Opcode_Strval");
    };
}
